src/main/scala/org/broadinstitute/dsde/firecloud/service/TrialService.scala

```diff
@@ -30,26 +35,130 @@ object TrialService {
-    case TerminateUser(userInfo) => terminateUser(userInfo) pipeTo sender
+    case TerminateUsers(userInfo, users) => terminateUsers(userInfo, users) pipeTo sender
   }
 
-  // TODO: implement fully! Check that the user does not already have a state, before overwriting.
-  // this method exists solely for developer-testing purposes right now.
-  private def enableUser(userInfo: UserInfo): Future[PerRequestMessage] = {
-    // build the state that we want to persist to indicate the user is enabled
-    val now = Instant.now
-    val zero = Instant.ofEpochMilli(0)
-    val enabledStatus = UserTrialStatus(userInfo.id, Some(TrialStates.Enabled), now, zero, zero, zero)
-    thurloeDao.saveTrialStatus(userInfo, enabledStatus) map { _ =>
-      RequestComplete(spray.http.StatusCodes.OK)
+  private def enableUsers(managerInfo: UserInfo, users: Seq[String]): Future[PerRequestMessage] = {
+    // TODO: Handle multiple users
+    require(users.size == 1, "For the time being, we can only enable one user at a time.")
+
+    val user = users.head
+
+    // Define what to overwrite in user's status
+    val statusTransition: UserTrialStatus => UserTrialStatus =
+      currentStatus =>
+        currentStatus.copy(
+          userId = managerInfo.id,
+          state = Some(Enabled),
+          enabledDate = Instant.now
+        )
+
+    updateUserState(managerInfo, user, statusTransition)
+  }
+
+  private def disableUsers(managerInfo: UserInfo, users: Seq[String]): Future[PerRequestMessage] = {
+    // TODO: Handle multiple users
+    require(users.size == 1, "For the time being, we can only disable one user at a time.")
+
+    val user = users.head
+
+    // Define what to overwrite in user's status
+    val statusTransition: UserTrialStatus => UserTrialStatus =
+      currentStatus =>
+        currentStatus.copy(
+          userId = managerInfo.id,
+          state = Some(Disabled)
+        )
+
+    updateUserState(managerInfo, user, statusTransition)
+  }
+
+  private def terminateUsers(managerInfo: UserInfo, users: Seq[String]): Future[PerRequestMessage] = {
+    // TODO: Handle multiple users
+    require(users.size == 1, "For the time being, we can only terminate one user at a time.")
+
+    val user = users.head
+
+    // Define what to overwrite in user's status
+    val statusTransition: UserTrialStatus => UserTrialStatus =
+      currentStatus =>
+        currentStatus.copy(
+          userId = managerInfo.id,
+          state = Some(Terminated),
+          terminatedDate = Instant.now
+        )
+
+    updateUserState(managerInfo, user, statusTransition)
+  }
+
+  private def updateUserState(managerInfo: UserInfo,
+                              user: String,
+                              statusTransition: UserTrialStatus => UserTrialStatus): Future[PerRequestMessage] = {
+    // TODO: Handle unregistered users which get 404 from Sam causing adminGetUserByEmail to throw
+    // TODO: Handle errors that may come up while querying Sam
```

Comment:
You could do something like this to handle all SAM-related errors, and add more specific error messages (i.e. 404's):
```
    samDao.adminGetUserByEmail(RawlsUserEmail(user)).flatMap { regInfo =>
      val workbenchUserInfo = WorkbenchUserInfo(regInfo.userInfo.userSubjectId, regInfo.userInfo.userEmail)
      updateTrialStatus(managerInfo, workbenchUserInfo, statusTransition).map {
        case StatusUpdate.Success => RequestComplete(NoContent)
        case StatusUpdate.Failure => RequestComplete(InternalServerError)
      }
    }.recover {
      case f: FireCloudExceptionWithErrorReport if f.errorReport.statusCode.isDefined =>
        RequestComplete(f.errorReport.statusCode.get)
      case t: Throwable =>
        throw new FireCloudExceptionWithErrorReport(ErrorReport.apply(StatusCodes.InternalServerError, t))
    }
```

Reply:
Thanks for the suggestion! Hoping to do something like that in a follow-on PR.